cmd/dockerfile/digests.go

```diff
@@ -0,0 +1,119 @@
+package dockerfile
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/buildsafedev/bsf/cmd/styles"
+	"github.com/google/go-containerregistry/pkg/crane"
+	"github.com/spf13/cobra"
+)
+
+var DGCmd = &cobra.Command{
+	Use:     "digests",
+	Short:   "Replace Dockerfile image tags with immutable digests",
+	Aliases: []string{"dg"},
+	Run: func(cmd *cobra.Command, args []string) {
+		if len(args) < 1 {
+			fmt.Println(styles.HintStyle.Render("hint:", "run `bsf dockerfile digests <Dockerfile>` to replace image tags with digests"))
+			os.Exit(1)
+		}
+
+		dockerfile := args[0]
+		file, err := os.ReadFile(dockerfile)
+		if err != nil {
+			fmt.Printf("Error opening Dockerfile: %v\n", err)
+			os.Exit(1)
+		}
+
+		var dgMap = map[string]string{}
+
+		line, err := readByte(file)
+		if err != nil {
+			fmt.Printf("Error in readByte %v\n", err)
+			os.Exit(1)
+		}
+
+		dgMap, err = getDigest(line)
+		if err != nil {
+			fmt.Printf("Error retrieving digest %v\n", err)
+			os.Exit(1)
+		}
+
+		updatedData, err := updateDockerfileWithDigests(file, dgMap)
+		if err != nil {
+			fmt.Printf("Error updating Dockerfile with digests: %v\n", err)
+			os.Exit(1)
+		}
+
+		if err := os.WriteFile(dockerfile, updatedData, 0644); err != nil {
+			fmt.Printf("Error writing updated Dockerfile: %v\n", err)
+			os.Exit(1)
+		}
+	},
+}
+
+func getDigest(lines []string) (map[string]string, error) {
+	var (
+		dgMap = make(map[string]string)
+		wg    sync.WaitGroup
+		mu    sync.Mutex
+	)
+
+	for _, line := range lines {
+		wg.Add(1)
+		go func(line string) {
+			defer wg.Done()
+			dg, err := crane.Digest(line)
+			if err != nil {
+				fmt.Println(styles.WarnStyle.Render("warning:", "skipping ", line, "can't find"))
+				return
+			}
+			mu.Lock()
+			dgMap[line] = dg
+			mu.Unlock()
+		}(line)
+	}
+
+	wg.Wait()
+	return dgMap, nil
+}
+
+func readByte(file []byte) ([]string, error) {
+	scanner := bufio.NewScanner(strings.NewReader(string(file)))
+	re := regexp.MustCompile(`FROM\s+(\S+):(\S+)`)
```

Comment:
will this work for let's say `FROM node:latest AS build`?

Reply:
bug here: I tried this and it's generating something like this:
`FROM sha256:c8a559f733bf1f9b3c1d05b97d9a9c7e5d3647c99abedaf5cdd3b54c9cbb8eff AS build`